### PR TITLE
helm chart failed with sidecar injector

### PR DESCRIPTION
**Pull request by linsun**
https://github.com/istio/istio/issues/2935, able to get a dry run working `helm install --dry-run --debug istio`, will work on clean up tmr.

cc @sdake @ayj

### CODE REVIEW

**Comment by sdake** [comment]
@linsun master(7ce862dc188fa01a5e27d84f90ef86b20c9f2807) works for me with Kubernetes 1.9.2 after running `updateVersion.sh -a $HUB,$TAG` where the images are built from the same commit.  Running updateVersion.sh is mandatory - or other problems are reported via Helm.  My kubeadm configuration without Mutating webhook is:
```
apiVersion: kubeadm.k8s.io/v1alpha1
kind: MasterConfiguration
apiServerExtraArgs:
  runtime-config: admissionregistration.k8s.io/v1beta1
  feature-gates: AllAlpha=true
pod-network-cidr: 192.168.0.0/16
```
I'm sorting out how to enable the mutating webhook now.

Is there a defined policy on a minimum version pin for Kubernetes for the installation tools?  It appears master has been modified to work with kubernetes>=1.9.0.  `updateVersion.sh` indicates that the configs are designed to work with Kubernetes 1.9.0+.  Some of the changes in this PR are changing from released stable versions to beta versions.  It would help me to understand if such a policy exists.

Cheers
-steve

**Comment by linsun** [comment]
So we support k8s 1.7, 1.8 and 1.9.  Only mutating web hook is supported in 1.9 as an alpha feature.  The default helm chart should be targeted for 1.7 or 1.8 with manual injection with a flag to allow users to enable mutating web hooks on 1.9 if they choose to.

**Comment by sdake** [comment]
@linsun thanks for the response.  Note the initializer worked for me in 1.8 without manual injection.  Does the new injection code work on 1.8, or only 1.9?

I have searched a bit for mutating web hook on google, and can't seem to find a definition of what it does in documentation.  Do you happen to have a link handy?

**Comment by andraxylia** [comment]
The mutating webhooks require k8s 1.9. Unfortunately, the initializers have been deprecated in 1.9, so we were forced to deprecate them as well. We will have to use manual injection for k8s < 1.9.

https://istio.io/docs/setup/kubernetes/sidecar-injection.html

**Comment by ayj** [comment]
See https://github.com/kubernetes/community/blob/master/contributors/design-proposals/api-machinery/admission-control-webhooks.md for background information on mutating webhooks.

https://kubernetes.io/docs/admin/extensible-admission-controllers/#external-admission-webhooks hasn't been updated yet for Mutating webhooks and the admissionregistration.k8s.io/v1beta1`  API group.

File: install/kubernetes/helm/istio/charts/sidecar-injector/templates/mutatingwebhook.yaml [hunk 1]
```diff
@@ -0,0 +1,29 @@
+{{- if .Values.global.sidecarInjectEnabled }}
```
[the remark this answers is not in the document]
Comment (sdake) [reply install/kubernetes/helm/istio/charts/sidecar-injector/templates/mutatingwebhook.yaml]: there is also a global.sidecar.enabled flag - which probably does the same thing as thing as this conditional.  I'm unclear if this helm chart has been tested on 1.9 with the webhooks (or simply 1.7).

**Comment by sdake** [comment]
We have several defects in this Helm chart as it stands.  I also think all of @linsun's work is an improvement given the policy of supporting kubernetes>=1.7.  Trying to fix all the various problems in one patch stream seems problematic (harder to review and test as there is no gating on the helm chart).

@linsun can you confirm if you have tested on 1.9?  If not, I'll take a look (thanks @ayj and @andraxylia for the pointers on injection).

**Comment by linsun** [comment]
Thank you @sdake - I don't have a 1.9 env, it would be great if you could take up on that.

I'll work on a gating test cc @luisyonaldo

**Comment by linsun** [comment]
Just make sure we are all on the same page, the goal is istio works on manual injection for 1.7 and 1.8, and manual & auto injection on 1.9+ while auto injection starts as alpha/dev feature for now till it is improved. cc @sdake @luisyonaldo @ayj

**Comment by ayj** [comment]
sgtm

**Comment by luisyonaldo** [comment]
Current logic check that you have sidecar-injector enabled and that you have mutation we hook api enabled. If not the only manual injection is possible. Haven't tested the auto injection yet.

File: install/kubernetes/helm/istio/charts/sidecar-injector/templates/mutatingwebhook.yaml (hunk 1 above)
[the remark this answers is not in the document]
Comment (linsun) [reply install/kubernetes/helm/istio/charts/sidecar-injector/templates/mutatingwebhook.yaml]: @ayj  I think that makes sense.  I'll do it in a separate PR.

**Pull request merged**